### DISCUSSION TITLE
How to add tool method into entity (like toMap) for converting Object

### DISCUSSION THREAD

**Discussion by darkomenx** [simolus3/drift#2585] [General]
Hello Drift,

**Context :** 
I would like to display my data (persisted into SQLite) into a Datatable Widget Flutter.
This Datatable Widget work around a DataRow Object that can be construct from a Map<String, Dynamic> object.
But the return object of my query Drift is a List<Entity>

**Idea:**
I would like to add a method into my entity (like toMap) to help me to convert object easily into a loop

**Issue:** 
If a add method into entity, this method is not recognize and I don't see it into generated drift_database.g.dart file.

**Questions :** 
1. How can I manage efficiently my return drift response from database to display this easily into datatable ?
2. How can I add method into entity that can be access into loop statement ?
3. Perhaps the unique solution and mandatory is use custom expressions queries and map function into it ?
4. If Drift can't directly manage added method into entity,  then how to create this concept into dart or flutter ?

_My code examples :_ 

**Query into repository :** 

```
  Future<List<Operation>> getOperations() async
  {
    List<Operation> allOperations = await db.select(db.operations).get();

    return allOperations;
  }
```

**treatment of return query :** 
```

      final List<Operation> operations = await _operationRepository.getOperations();

      for (final operation in operations)
      {
        operation.toMap() // toMap() method is not recognize here
      }
```

**My entity :** 

```
import 'package:drift/drift.dart';

@DataClassName('Operation')
class Operations extends Table
{
  IntColumn get id => integer().autoIncrement()();
  IntColumn get thirdParty => integer().nullable()();
  TextColumn get label => text().withLength(min: 1, max: 32)();
  IntColumn get category => integer().nullable()();

  Map<String, dynamic> toMap()
  {
    return {
      'id':         id,
      'thirdParty': thirdParty,
      'label':      label,
      'category':   category,
    };
  }
}
```

What do you think about those ?

Thank you for your package work and help.....

**Comment by simolus3** [comment]
Methods you put in the class describing the table don't show up in the class describing a row of that table. You can use [custom row classes](https://drift.simonbinder.eu/docs/advanced-features/custom_row_classes/) to customize them entirely. An easier workaround might be to do this:

```dart
extension RowToMap<D extends Insertable<D>> on D {
  Map<String, Object?> toMap() {
    return {
      for (final entry in toColumns(false).entries)
        entry.key: (entry.value as Variable).value,
    };
  }
}
```

That should allow you to call `toMap()` on all data classes generated by drift.

**Comment by darkomenx** [comment]
using Row class consist of adding annotation in entity class but I already have one and this is not possible to have 2 distinct annotations

If I change @DataClassName to @UseRowClass all my code base is invalidated immediatly.

**Reply by simolus3** [reply]
With `@UseRowClass`, the idea is that you write the data class instead of letting drift generate one for you. So the two annotations are incompatible with each other.

**Comment by darkomenx** [comment]
OK it seems that custom Row Classes is powerful (I more understand now).
For now, I prefer your extension way resolution than rowClass but I keep in mind this possibility.

My extension seems to be operate good 👍 

Why do not included this base and popular function toMap() directly into your generated drift code ?

Thank you for your help again.
